### DISCUSSION TITLE
prevent user creation on new OAuth2 auth ?

### DISCUSSION THREAD

**Discussion by tournesolo** [pocketbase/pocketbase#4357] [Q&A] [answered]
I need that only pre-approved users can sign into my pocketbase app, not anyone who has a google/facebook/etc account. But as stated in an older question:

> PocketBase will automatically create the user for you on new OAuth2 auth

How can I prevent that and instead check if the newly authenticated user is already part of the pre-approved users? Thanks, and thanks again for the wonderful pocketbase :)

**Comment by tournesolo** [comment]
Despite having searched for similar questions before posting, I found a seemingly relevant post 2 minutes after asking... So would this earlier answer be the the right solution?

```
onRecordBeforeCreateRequest((e) => {
    // allow for admins
    if (e.httpContext.get("admin")) {
        return
    }

    // disallow for everyone else
    throw new ForbiddenError("You are not allowed to perform this action.") 
}, "users")
```

Ie this should only allow admins to create new users, and hence block automatic user creation on oauth2 signin.

**Comment by ganigeorgiev** [comment, marked as answer]
If by "pre-approved user" you mean created by an admin, then you can "lock" the users collection Create API rule (aka. make it "Admin only") and the OAuth2 authentication for new/unlinked users will fail on its own.

Alternatively you can use the hooks as you found out. If you are targeting only the OAuth2 user create authentication then maybe a better hook would be the `OnRecordBeforeAuthWithOAuth2Request` hook:
- [Go](https://pocketbase.io/docs/go-event-hooks/#onrecordbeforeauthwithoauth2request):
    ```go
    app.OnRecordBeforeAuthWithOAuth2Request().Add(func(e *core.RecordAuthWithOAuth2Event) error {
        // nil Record means that it will try to automatically create one (if it can)
        if e.Record == nil {
            return apis.NewForbiddenError("You are not allowed to perform this request", nil)
        }

        return nil
    })
    ```

- [JS `pb_hooks`](https://pocketbase.io/docs/js-event-hooks/#onrecordbeforeauthwithoauth2request): 
    ```js
    onRecordBeforeAuthWithOAuth2Request((e) => {
        // empty Record means that it will try to automatically create one (if it can)
        if (!e.record) {
            throw new ForbiddenError("You are not allowed to perform this request")
        }
    })
    ```


With the next release there will be also `@request.source/context` field which will allow to apply different constraints based on from where the API rule is checked, for example you'll be able to define a Create API rule that allow regular user create requests but disallow user creation from an OAuth2 call:

```js
@request.context != "oauth2"
```

_edit: updated with the actual field name_

**Reply by eevleevs** [reply]
Should `@request.source` be available now? I am getting 

`Invalid rule. Raw error: invalid left operand "@request.source" - failed to resolve field "@request.source".`

from 0.23

**Reply by ganigeorgiev** [reply]
It is `@request.context` (the same apply for the older versions). See https://pocketbase.io/docs/api-rules-and-filters/#filters-syntax.

**Comment by tournesolo** [comment]
>  you can "lock" the users collection Create API rule (aka. make it "Admin only") and the OAuth2 authentication for new/unlinked users will fail on its own

That's an even more elegant solution, thanks! (Thanks also for pointing out `OnRecordBeforeAuthWithOAuth2Request`)
And yes, the app only has oauth2 login, no local username/password users.

**Comment by tournesolo** [comment]
After quickly trying it out, a couple of drawbacks:

- locking user creation results in a generic (cryptic) error ("ClientResponseError 400: Failed to authenticate."), whereas onRecordBeforeCreateRequest / OnRecordBeforeAuthWithOAuth2Request can presumably be made to return something that allows the frontend to give an informative error to the user
-  in all variants, the error / refusal happens _after_ the user has already approved the pocketbase app in their auth provider, so now they have a useless dangling approval.

It would thus seem nicer for me to find a way to check that the user is legit before they start the oauth2 song and dance (ideally without asking for their email/id twice, once to pre-check and once during authorisation at their provider).

**Reply by ganigeorgiev** [reply]
> seem nicer for me to find a way to check that the user is legit 

I don't know how this can be achieved as you'll need to get the user identifier (id/email/etc.) from somewhere and I don't think there is a way to fetch the OAuth2 user info without the user to give you explicit OAuth2 app permission to read it.

**Reply by tournesolo** [reply]
The only 'nicer' way I can think of, though perhaps not feasible: ask the user for their id/email before sending them to the oauth2 provider, and if it is a legit user, send them to the oauth provider along with the id/email they provided, so that the oauth provider doesn't have to reask for that piece of info. But I guess this is getting too complex / off-track.
